skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import operator
+from functools import wraps
+
+from ._intersection import IntervalTree
+from skbio.util._decorator import experimental
+
+
+class Interval:
+    """Stores the location and metadata of an interval feature.
+
+    This class stores an interval feature. An interval feature
+    is defined as a sub-region of a sequence that is a functional
+    entity, e.g., a gene, a riboswitch, an exon, etc. It can span
+    a single contiguous region or multiple non-contiguous regions (e.g.
+    multiple exons in a transcript, or multiple genes in an operon).
+
+    Parameters
+    ----------
+    interval_metadata : object
+        A reference to the ``IntervalMetadata`` object that this
+        ``Interval`` object is associated to.
+    locations : iterable of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : iterable of tuple of bool
+        List of tuples, representing the openness of each location coordinates.
+        If this isn't specified, then all of the boundaries are True. If any
+        of the coordinate boundaries is False, it indicates that the exact
+        boundary point of a interval feature is unknown. The location may
+        begin or end at some points outside the specified coordinates. This
+        accommodates the location format [1]_ of INSDC.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as "strand", "gene_name", or "product".
+
+    Attributes
+    ----------
+    locations
+    boundaries
+    metadata
+
+    See Also
+    --------
+    skbio.metadata.IntervalMetadata
+
+    Examples
+    --------
+    >>> from skbio.metadata import Interval, IntervalMetadata
+    >>> gene = Interval(interval_metadata=IntervalMetadata(),
```

Comment:
It might be clearer if the `IntervalMetadata()` call was not inlined as it will match normal usage that way.
```python
>>> interval_metadata = IntervalMetadata()
>>> gene = Interval(interval_metadata, locations=...)
```